internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+type RemoteRecoverableError struct {
+	error
+}
+
+func (e RemoteRecoverableError) Error() string {
+	if e.error != nil {
+		return e.Error()
+	} else {
+		return ""
+	}
+}
+
+// WriteClient
+// an interface used for testing
+//go:generate mockgen -package=metrics -destination=write_client_mock.go . WriteClient
+type WriteClient interface {
+	Write(context.Context, []byte) error
+}
+
+type writeClient struct {
+	client remote.WriteClient
+}
+
+func (w *writeClient) Write(ctx context.Context, data []byte) error {
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
+	serverTimeout, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_SERVER_TIMEOUT"))
+	if serverTimeout == 0 {
+		serverTimeout = time.Second * 10
+	}
+
+	// create configuration
+	r.config = RemoteWriteConfiguration{
+		Url:     serverURL,
+		Timeout: serverTimeout,
+	}
+
+	if r.config.Url == "" { // feature is disabled
+		return nil
+	}
+
+	// create client
+	if r.Client == nil {
+		serverURL, err := url.Parse(r.config.Url)
+		if err != nil {
+			return err
+		}
+		client, err := remote.NewWriteClient(r.deviceID, &remote.ClientConfig{
+			Timeout: model.Duration(r.config.Timeout),
+			URL: &config_util.URL{
+				URL: serverURL,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		r.Client = &writeClient{
+			client: client,
+		}
+	}
+
+	return nil
+}
+
+func (r *RemoteWrite) Init(config models.DeviceConfigurationMessage) error {
+	err := r.Update(config)
+	go r.writeRoutine()
+	return err
+}
+
+// writeRoutine
+// Used as the goroutine function for handling ongoing writes to remote server
+func (r *RemoteWrite) writeRoutine() {
+	log.Infof("metric remote writer %s started", r.deviceID)
+	for {
+		r.Write()
+		time.Sleep(r.waitInterval)
+	}
+}
+
+// write
+// Writes all the metrics. Stops when either no more metrics or failure (after retries)
+func (r *RemoteWrite) Write() {
+
+	if r.Client == nil { // feature disabled
+		return
+	}
+
+	// start writing
+	for maxTSDB := r.tsdb.MaxTime(); maxTSDB.Sub(r.LastWrite) > 0; maxTSDB = r.tsdb.MaxTime() {
+
+		// set range start and end
+		minTSDB, err := r.tsdb.MinTime()
+		if err != nil {
+			log.Error("failed reading TSDB min", err)
+			return
+		}
+
+		rangeStart := r.LastWrite.Add(time.Millisecond) // TSDB time is in milliseconds
+		if rangeStart.Sub(minTSDB) < 0 {
+			rangeStart = minTSDB
+		}
+		rangeEnd := rangeStart.Add(r.RangeDuration)
+		if rangeEnd.Sub(maxTSDB) > 0 {
+			rangeEnd = maxTSDB
+		}
+
+		log.Infof("going to write metrics range %s(%d)-%s(%d). TSDB min max: %s(%d)-%s(%d)",
```

Comment:
Maybe a debug level is more appropriate here?

Reply:
On one hand yes. On the other if we see that metrics are missing, how can we tell that the device sent them?  If no one else objects and you still think it should be debug then I'll be glad to change (less noise in log)

Reply:
@pkliczewski @jakub-dzon @eloycoto @masayag what do u think? Gloria raised an important concern

Reply:
I'm ok to get it on info.

Reply:
@ydayagi Wouldn't logging only the failure be enough to debug this issue?
I mean if it reaches to line 215,  the device is about to write the data.
If it failed, there should be an error to record that.